libs/utils/src/tabulation.ts

```diff
@@ -408,33 +408,52 @@ export function groupMapToGroupList<T>(
   return list;
 }
 
+/**
+ * Yield to the event loop during long-running tabulation operations.
+ */
+export async function yieldToEventLoop(): Promise<void> {
+  await new Promise((resolve) => {
+    setImmediate(resolve);
+  });
+}
+
 /**
  * Tabulates iterable cast vote records into election results, grouped by
  * the attributes specified {@link Tabulation.GroupBy} parameter.
  */
-export function tabulateCastVoteRecords({
+export async function tabulateCastVoteRecords({
```

Comment:
Ok, it looks to me like mixed `async` and `sync` definitions are not possible but mixed `Promise` and non-`Promise` return values are possible. So I would need to find a way to write this without `await`.

In the method I changed to `async`, I'm `await`ing `yieldToEventLoop` at every `N` CVRs. How could I do this with promises? Might that have to be recursive, and then risk too much stack depth?

Reply:
The simplest way is to just move as much of the code out into a helper function as possible and then call it from duplicated sync and async versions. And yeah, probably just "recursively" call via `setImmediate` or similar. Since it's not actually a recursive call I don't think you'll hit stack trace issues but that's easy enough to test.

Reply:
Hmmm.

re: Separating, I think this logic will be a big challenging to separate much further (because it has two branches, each of them looping differently in sync vs. async).

re: Recursive, I don't see how to write this function without it _actually_ being recursive, because I'm imagining you'd set up some outer scope state and then an inner helper function which returns a promise or non-promise within it, the promises being created recursively and advancing tabulation incrementally. But this is probably just limitations of my understanding.

In any case, discussing this has made me think that the one frontend tradeoff of having to working around this for rendering test decks is worth it, so I'm planning to just merge it as-is unless that sounds wrong.

Reply:
Oh, I couldn't imagine how this would not be recursive but I see now how it works! I really appreciate the explanation.